File: src/com/javascript-practice/binarytree/depthFirstValues.js
/**
 * Write a function, depthFirstValues, that takes in the root of a binary tree. 
 * The function should return an array containing all values of the tree in 
 * depth-first order.
 * 
 * 
 * n = number of nodes
Time: O(n)
Space: O(n)
for both recursion and iterative
 */
class Node {
constructor(val) {
    this.val = val;
    this.left = null;
    this.right = null
    }
}

const depthFirstValues = (root) => {
    if(root == null) return [];
    let values = [];
    let stack = [root];
    while(stack.length > 0) {
        let current  = stack.pop();
        values.push(current.val);
        if(current.right !== null) {
            stack.push(current.right);
        }
        if(current.left !== null) {
            stack.push(current.left);
        }
        
    }
    return values;
}


const a = new Node('a');
const b = new Node('b');
const c = new Node('c');
const d = new Node('d');
const e = new Node('e');
const f = new Node('f');

a.left = b;
a.right = c;
b.left = d;
b.right = e;
c.right = f;

//      a
//    /   \
//   b     c
//  / \     \
// d   e     f
/**
 * //   a
//    /   \
//   b     c
//  / \     \
// d   e     f

depthFirstValues(a); 
//    -> ['a', 'b', 'd', 'e', 'c', 'f']
 */

let values = depthFirstValues(a);
//console.log(values);

const depthFirstValueRecurse = (root) =>  {
    if(root == null)  return []
    const leftValues = depthFirstValueRecurse(root.left); // [b,d,e]
   // const rightValues =  depthFirstValueRecurse(root.right); // [c,f]
   // result = [...leftValues];

    return [root.val, ...leftValues];
}
let recurseSol = depthFirstValueRecurse(a);
console.log(recurseSol);

File: src/com/structy/binaryTree/DepthFirstValues.java
package com.structy.binaryTree;

import java.util.ArrayDeque;
import java.util.ArrayList;
import java.util.Deque;
import java.util.List;

public class DepthFirstValues {

    public List<Integer> depthFirstValues(Node root) {
        return dfs(root, new ArrayList<Integer>());
    }
    private List<Integer> dfs(Node root, List<Integer> result){
        if(root == null){
            return result;
        }
         result.add(root.val);
        //if(root.left!=null){
          List<Integer> leftValues =  dfs(root.left, result);
       // }
        //if(root.right!=null){
           List<Integer> rightValues =  dfs(root.right, result);
       // }
        // result.add(root.val);
        //result.addAll(leftValues);
        //result.addAll(rightValues);
        return result;
    }
    public List<Integer> depthFirstValuesIterative(Node root) {
        if(root == null) return new ArrayList<>();
        Deque<Node> stack   = new ArrayDeque<>();
        stack.push(root);
        List<Integer> result = new ArrayList<>();
        while (stack.size() > 0){
            Node current = stack.pop();
            result.add(current.val);
            if(current.right!=null)
                stack.push(current.right);
            if(current.left!=null)
                stack.push(current.left);
        }
        return result;
    }
    public static void main(String[] args) {
        Node a = new Node(1);
        Node b = new Node(2);
        Node c = new Node(3);
        Node d = new Node(4);
        Node e = new Node(5);
        Node f = new Node(6);

        a.left = b;
        a.right = c;
        b.left = d;
        b.right = e;
        c.right = f;

//      1
//    /   \
//   2     3
//  / \     \
// 4   5     6

        DepthFirstValues depthFirstValues = new DepthFirstValues();
        System.out.println(depthFirstValues.depthFirstValues(a));
        System.out.println(depthFirstValues.depthFirstValuesIterative(a));
    }
}
